File: src/store/sections.ts
// Byte-exact sectioning of a raw API body (TRDD-K3WDPR7M Phase 2).
//
// WHY SECTION AT ALL: every LLM request re-sends the WHOLE conversation plus an identical ~342 KB
// `tools` array. Measured over 40 real bodies: 27.9 MB raw -> 2,069 unique sections. The bodies are
// ~86% pure repetition, so content-addressing the SECTIONS makes a repeat turn cost almost nothing
// (only the parts that actually changed are new).
//
// WHY BYTE OFFSETS AND NOT JSON.parse -> JSON.stringify: reconstruction must be BYTE-IDENTICAL to the
// source file, and a re-serialization round-trip only *happens* to be byte-exact — it rests on V8's
// key-order rules, its number formatting, and its string escaping all matching the producer's. That
// held for 25/25 sample bodies, which is exactly the kind of evidence that lulls you into shipping a
// corpus you cannot restore. So we never re-serialize: we slice the ORIGINAL text into spans and
// concatenate them back. Byte-identity is then true BY CONSTRUCTION, not by luck.
import { createHash } from 'crypto'

/** A span smaller than this stays inline as literal glue: a content-addressed row costs more (hash +
 *  index + row overhead) than it can ever save on a 20-byte scalar. Tuned to keep `model`, `stream`,
 *  `max_tokens` etc. inline while every message block and tool definition gets deduped. */
export const MIN_BLOB_BYTES = 128

/** One piece of a body. Concatenating every part's bytes, in order, reproduces the source EXACTLY. */
export type Part =
  /** Literal source bytes: punctuation, keys, and any value too small to be worth deduping. */
  | { kind: 'lit'; text: string }
  /** A content-addressed span. `path`/`idx`/`n` are the IN-CLEAR index the user asked for: they answer
   *  "what is in this context / what changed / what is growing" with ZERO decompression. */
  | { kind: 'blob'; sha: string; n: number; path: string; idx: number }

export interface Sectioned {
  parts: Part[]
  /** sha256 -> the exact span text. The caller writes these content-addressed (a repeat costs 0). */
  blobs: Map<string, string>
}

export function sha256(text: string): string {
  return createHash('sha256').update(Buffer.from(text, 'utf8')).digest('hex')
}

/** Skip JSON whitespace. */
function ws(s: string, i: number): number {
  while (i < s.length && (s[i] === ' ' || s[i] === '\n' || s[i] === '\r' || s[i] === '\t')) i++
  return i
}

/**
 * Return the index just past the JSON value starting at `i`. A minimal scanner — it does NOT validate
 * the document (the producer is Claude Code, and a malformed body would fail the sha256 gate anyway);
 * it only needs to find the value's EXACT extent so we can slice it.
 *
 * The string case is the one that must be exactly right: a `"` inside a string, or a `{`/`[` inside
 * one, would otherwise desynchronize the whole scan. `\\` must be consumed as a unit, or a trailing
 * backslash before the closing quote (`"C:\\"`) would eat the quote and run off the end.
 */
export function scanValue(s: string, i: number): number {
  i = ws(s, i)
  const c = s[i]
  if (c === '"') {
    i++
    while (i < s.length) {
      if (s[i] === '\\') { i += 2; continue } // an escape pair is ONE unit — never look at its second char
      if (s[i] === '"') return i + 1
      i++
    }
    throw new Error('unterminated string in JSON body')
  }
  if (c === '{' || c === '[') {
    const open = c
    const close = c === '{' ? '}' : ']'
    let depth = 0
    while (i < s.length) {
      const ch = s[i]
      if (ch === '"') { i = scanValue(s, i); continue } // strings are opaque — braces inside don't count
      if (ch === open) depth++
      else if (ch === close) { depth--; if (depth === 0) return i + 1 }
      i++
    }
    throw new Error('unterminated object/array in JSON body')
  }
  // number | true | false | null — run to the next structural character.
  const start = i
  while (i < s.length && !',}] \n\r\t'.includes(s[i])) i++
  if (i === start) throw new Error(`unexpected character ${JSON.stringify(c)} at ${i}`)
  return i
}

/** Emit a span as a blob part (deduped) or as literal glue when it is too small to be worth a row. */
function emit(out: Sectioned, span: string, path: string, idx: number): void {
  const n = Buffer.byteLength(span, 'utf8')
  if (n < MIN_BLOB_BYTES) {
    out.parts.push({ kind: 'lit', text: span })
    return
  }
  const sha = sha256(span)
  out.blobs.set(sha, span) // a Map keyed by content IS the dedup — a repeat overwrites itself
  out.parts.push({ kind: 'blob', sha, n, path, idx })
}

/**
 * Split a raw body into parts. Top-level keys become sections; a top-level ARRAY is split
 * ELEMENT-WISE.
 *
 * The element-wise split is the load-bearing part and it must be GENERIC, not messages-only: the
 * `tools` array is 342 KB and byte-identical on EVERY request (172 items). A messages-only sectioner
 * would re-store those 342 KB every single turn and leave most of the redundancy on the table.
 */
export function sectionize(raw: string): Sectioned {
  const out: Sectioned = { parts: [], blobs: new Map() }
  let i = ws(raw, 0)
  if (raw[i] !== '{') {
    // Not a top-level object (shouldn't happen for an API body) — store it whole rather than guess.
    emit(out, raw, '$', -1)
    return out
  }

  let lit = i // start of the pending literal run (glue we copy through verbatim)
  i++ // past '{'
  for (;;) {
    i = ws(raw, i)
    if (raw[i] === '}') { out.parts.push({ kind: 'lit', text: raw.slice(lit) }); break }
    if (raw[i] === ',') { i++; continue }

    const keyEnd = scanValue(raw, i)          // the key string, quotes included
    const key = JSON.parse(raw.slice(i, keyEnd)) as string
    let j = ws(raw, keyEnd)
    if (raw[j] !== ':') throw new Error(`expected ':' after key ${key}`)
    j = ws(raw, j + 1)

    if (raw[j] === '[') {
      // Split the array element-wise. Everything that is NOT an element — '[', ',', ']', and any
      // whitespace between them — is copied through as literal glue by slicing the source, so the
      // original formatting survives exactly. (Emitting a synthesized ',' instead would drop the
      // surrounding whitespace: real API bodies are compact so it would never show, which is
      // precisely the kind of "happens to work" that makes a corpus unrecoverable the one day it
      // doesn't.)
      out.parts.push({ kind: 'lit', text: raw.slice(lit, j + 1) })
      let k = j + 1
      let glue = k // start of the pending glue run INSIDE the array
      let elem = 0
      for (;;) {
        const p = ws(raw, k)
        if (raw[p] === ']') { k = p + 1; break }   // trailing ws + ']' belong to the next glue run
        if (raw[p] === ',') { k = p + 1; continue } // stays in the glue run — do not synthesize it
        const end = scanValue(raw, p)
        out.parts.push({ kind: 'lit', text: raw.slice(glue, p) })
        emit(out, raw.slice(p, end), key, elem++)
        glue = end
        k = end
      }
      lit = glue // from the end of the last element (its trailing ws + ']') onward
      i = k
      continue
    }

    // A scalar or object value: glue up to it stays literal, the value itself is content-addressed.
    const end = scanValue(raw, j)
    out.parts.push({ kind: 'lit', text: raw.slice(lit, j) })
    emit(out, raw.slice(j, end), key, -1) // idx -1 = "not an array element"
    lit = end
    i = end
  }
  return out
}

/**
 * Rebuild the exact source text. `getBlob` resolves a sha256 to its span; a MISSING blob THROWS —
 * silently emitting a gap would hand back a corrupt body that still looks like JSON, and the caller
 * would persist that as truth. Fail loudly instead (the sha256 gate would catch it, but only after
 * the damage is on disk).
 */
export function reassemble(parts: Part[], getBlob: (sha: string) => string | undefined): string {
  let out = ''
  for (const p of parts) {
    if (p.kind === 'lit') { out += p.text; continue }
    const span = getBlob(p.sha)
    if (span === undefined) throw new Error(`missing blob ${p.sha} (${p.path}[${p.idx}], ${p.n} bytes)`)
    out += span
  }
  return out
}

File: rust-core/crates/agentlens-store/src/sections.rs
//! Byte-exact sectioning of a raw API body — port of `src/store/sections.ts`.
//!
//! WHY BYTE OFFSETS AND NOT parse→re-serialize: reconstruction must be BYTE-IDENTICAL to the
//! source, and a re-serialization round trip only *happens* to be byte-exact. We slice the
//! ORIGINAL text and concatenate it back — identity by construction, not by luck. Rust scans
//! BYTES (all structural JSON characters are ASCII, so multibyte UTF-8 can never collide),
//! which is naturally byte-exact where the TS scanner had to be careful with UTF-16 indices.

use sha2::{Digest, Sha256};

/// A span smaller than this stays inline as literal glue (a content-addressed row costs more
/// than it saves on a 20-byte scalar). Same constant as the TS sectioner.
pub const MIN_BLOB_BYTES: usize = 128;

#[derive(Debug, Clone, PartialEq)]
pub enum Part {
    Lit { text: String },
    Blob { sha: String, n: u64, path: String, idx: i64 },
}

pub struct Sectioned {
    pub parts: Vec<Part>,
    /// sha256 → exact span text. Keyed by content — the map IS the dedup.
    pub blobs: std::collections::HashMap<String, String>,
    /// Blob insertion order (the TS Map preserves it; HashMap does not) — the appender writes
    /// blobs in this order so cross-engine part files stay comparable.
    pub blob_order: Vec<String>,
}

pub fn sha256_hex(text: &str) -> String {
    let mut h = Sha256::new();
    h.update(text.as_bytes());
    let out = h.finalize();
    let mut s = String::with_capacity(64);
    for b in out {
        use std::fmt::Write;
        let _ = write!(s, "{b:02x}");
    }
    s
}

fn ws(s: &[u8], mut i: usize) -> usize {
    while i < s.len() && matches!(s[i], b' ' | b'\n' | b'\r' | b'\t') {
        i += 1;
    }
    i
}

/// Index just past the JSON value starting at `i`. Same minimal scanner as the TS original:
/// escape pairs consume as ONE unit; strings are opaque to brace counting.
pub fn scan_value(s: &[u8], mut i: usize) -> Result<usize, String> {
    i = ws(s, i);
    let Some(&c) = s.get(i) else { return Err(format!("unexpected end of body at {i}")) };
    if c == b'"' {
        i += 1;
        while i < s.len() {
            match s[i] {
                b'\\' => i += 2, // an escape pair is ONE unit — never look at its second byte
                b'"' => return Ok(i + 1),
                _ => i += 1,
            }
        }
        return Err("unterminated string in JSON body".into());
    }
    if c == b'{' || c == b'[' {
        let (open, close) = if c == b'{' { (b'{', b'}') } else { (b'[', b']') };
        let mut depth = 0usize;
        while i < s.len() {
            let ch = s[i];
            if ch == b'"' {
                i = scan_value(s, i)?;
                continue;
            }
            if ch == open {
                depth += 1;
            } else if ch == close {
                depth -= 1;
                if depth == 0 {
                    return Ok(i + 1);
                }
            }
            i += 1;
        }
        return Err("unterminated object/array in JSON body".into());
    }
    let start = i;
    while i < s.len() && !matches!(s[i], b',' | b'}' | b']' | b' ' | b'\n' | b'\r' | b'\t') {
        i += 1;
    }
    if i == start {
        return Err(format!("unexpected character {:?} at {i}", c as char));
    }
    Ok(i)
}

fn emit(out: &mut Sectioned, span: &str, path: &str, idx: i64) {
    let n = span.len();
    if n < MIN_BLOB_BYTES {
        out.parts.push(Part::Lit { text: span.to_owned() });
        return;
    }
    let sha = sha256_hex(span);
    if !out.blobs.contains_key(&sha) {
        out.blob_order.push(sha.clone());
    }
    out.blobs.insert(sha.clone(), span.to_owned());
    out.parts.push(Part::Blob { sha, n: n as u64, path: path.to_owned(), idx });
}

/// Split a raw body into parts — top-level keys become sections, a top-level array splits
/// element-wise, and EVERY non-value byte (punctuation, whitespace) survives as literal glue.
pub fn sectionize(raw: &str) -> Result<Sectioned, String> {
    let s = raw.as_bytes();
    let mut out = Sectioned { parts: Vec::new(), blobs: std::collections::HashMap::new(), blob_order: Vec::new() };
    let mut i = ws(s, 0);
    if s.get(i) != Some(&b'{') {
        emit(&mut out, raw, "$", -1);
        return Ok(out);
    }
    let mut lit = i; // start of the pending literal run
    i += 1; // past '{'
    loop {
        i = ws(s, i);
        match s.get(i) {
            Some(b'}') => {
                out.parts.push(Part::Lit { text: raw[lit..].to_owned() });
                break;
            }
            Some(b',') => {
                i += 1;
                continue;
            }
            None => return Err("unterminated top-level object".into()),
            _ => {}
        }
        let key_end = scan_value(s, i)?;
        let key: String = serde_json::from_str(&raw[i..key_end]).map_err(|e| format!("bad key: {e}"))?;
        let mut j = ws(s, key_end);
        if s.get(j) != Some(&b':') {
            return Err(format!("expected ':' after key {key}"));
        }
        j = ws(s, j + 1);

        if s.get(j) == Some(&b'[') {
            out.parts.push(Part::Lit { text: raw[lit..j + 1].to_owned() });
            let mut k = j + 1;
            let mut glue = k;
            let mut elem: i64 = 0;
            loop {
                let p = ws(s, k);
                match s.get(p) {
                    Some(b']') => {
                        k = p + 1;
                        break;
                    }
                    Some(b',') => {
                        k = p + 1;
                        continue;
                    }
                    None => return Err("unterminated array".into()),
                    _ => {}
                }
                let end = scan_value(s, p)?;
                out.parts.push(Part::Lit { text: raw[glue..p].to_owned() });
                emit(&mut out, &raw[p..end], &key, elem);
                elem += 1;
                glue = end;
                k = end;
            }
            lit = glue; // from the end of the last element (trailing ws + ']') onward
            i = k;
            continue;
        }

        let end = scan_value(s, j)?;
        out.parts.push(Part::Lit { text: raw[lit..j].to_owned() });
        emit(&mut out, &raw[j..end], &key, -1);
        lit = end;
        i = end;
    }
    Ok(out)
}

/// Rebuild the exact source text. A missing blob is an ERROR, never a silent gap.
pub fn reassemble(parts: &[Part], get_blob: impl Fn(&str) -> Option<String>) -> Result<String, String> {
    let mut out = String::new();
    for p in parts {
        match p {
            Part::Lit { text } => out.push_str(text),
            Part::Blob { sha, n, path, idx } => {
                let span = get_blob(sha).ok_or_else(|| format!("missing blob {sha} ({path}[{idx}], {n} bytes)"))?;
                out.push_str(&span);
            }
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sectionize_round_trips_byte_identically() {
        let big = "x".repeat(200);
        let raw = format!(
            "{{\"model\":\"claude-opus-5\",\"messages\":[{{\"role\":\"user\",\"content\":\"{big}\"}} , {{\"role\":\"assistant\",\"content\":\"{big}zzz\"}}],\"max_tokens\":5}}"
        );
        let sec = sectionize(&raw).expect("sectionize");
        let back = reassemble(&sec.parts, |sha| sec.blobs.get(sha).cloned()).expect("reassemble");
        assert_eq!(back, raw, "byte identity by construction");
        assert!(sec.blobs.len() >= 2, "both big elements content-address");
    }

    #[test]
    fn escapes_and_nested_braces_inside_strings_do_not_desync() {
        let pad = "p".repeat(150);
        let raw = format!(
            "{{\"a\":\"quote \\\" brace {{ bracket [ backslash \\\\ end {pad}\",\"b\":[1,2,3]}}"
        );
        let sec = sectionize(&raw).expect("sectionize");
        let back = reassemble(&sec.parts, |sha| sec.blobs.get(sha).cloned()).expect("reassemble");
        assert_eq!(back, raw);
    }

    #[test]
    fn non_object_body_stores_whole() {
        let raw = "[1,2,3]";
        let sec = sectionize(raw).expect("sectionize");
        let back = reassemble(&sec.parts, |sha| sec.blobs.get(sha).cloned()).expect("reassemble");
        assert_eq!(back, raw);
    }
}
